translator/trace/zipkin/protospan_to_zipkinv1.go

```diff
@@ -295,29 +292,78 @@ func OCSpanDataToZipkin(
 	}
 
 	// construct Annotations from s.Annotations and s.MessageEvents.
-	if len(s.Annotations) != 0 || len(s.MessageEvents) != 0 {
-		z.Annotations = make([]zipkinmodel.Annotation, 0, len(s.Annotations)+len(s.MessageEvents))
-		for _, a := range s.Annotations {
-			z.Annotations = append(z.Annotations, zipkinmodel.Annotation{
-				Timestamp: a.Time,
-				Value:     a.Message,
-			})
+	z.Annotations = timeEventsToSpanAnnotations(s.TimeEvents)
+
+	return z, nil
+}
+
+func resourceToSpanAttributes(resource *resourcepb.Resource, tags map[string]string) {
+	if resource != nil {
+		for key, value := range resource.Labels {
+			tags[key] = value
+		}
+	}
+}
+
+func timeEventsToSpanAnnotations(timeEvents *tracepb.Span_TimeEvents) []zipkinmodel.Annotation {
+	var annotations []zipkinmodel.Annotation
+	// construct Annotations from s.Annotations and s.MessageEvents.
+	if timeEvents != nil && len(timeEvents.TimeEvent) != 0 {
+		annotations = make([]zipkinmodel.Annotation, 0, len(timeEvents.TimeEvent))
+		for _, te := range timeEvents.TimeEvent {
+			if te == nil || te.Value == nil {
+				continue
+			}
+
+			switch tme := te.Value.(type) {
+			case *tracepb.Span_TimeEvent_Annotation_:
+				annotations = append(annotations, zipkinmodel.Annotation{
+					Timestamp: internal.TimestampToTime(te.Time),
+					Value:     derefTruncatableString(tme.Annotation.Description),
+				})
+			case *tracepb.Span_TimeEvent_MessageEvent_:
+				a := zipkinmodel.Annotation{
+					Timestamp: internal.TimestampToTime(te.Time),
+					Value:     tme.MessageEvent.Type.String(),
+				}
+				annotations = append(annotations, a)
+
+			}
 		}
-		for _, m := range s.MessageEvents {
-			a := zipkinmodel.Annotation{
-				Timestamp: m.Time,
+	}
+	return annotations
+}
+
+func attributesToSpanTags(attrMap map[string]*tracepb.AttributeValue, redundantKeys map[string]bool, tags map[string]string) {
+	if len(attrMap) != 0 {
+		for key, value := range attrMap {
+			if redundantKeys[key] {
+				// Already represented by something other than an attribute,
+				// skip it.
+				continue
 			}
-			switch m.EventType {
-			case trace.MessageEventTypeSent:
-				a.Value = "SENT"
-			case trace.MessageEventTypeRecv:
-				a.Value = "RECV"
-			default:
-				a.Value = "<?>"
+
+			switch v := value.Value.(type) {
```

Comment:
Consider adding a default case (maybe not needed if sure the proto won't be added to)

Reply:
This code will soon be replaced to use pdata. But I did it in 2 steps because it was some legacy here to use SpanData. Also OC is in maintenance so no changes expected there.